pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -288,8 +287,8 @@ public void readMoreEntries() {
                     return;
                 }
 
-                Set<PositionImpl> messagesToReplayNow = messagesToReplay.items(messagesToRead).stream()
-                        .map(pair -> new PositionImpl(pair.first, pair.second)).collect(toSet());
+                Set<PositionImpl> messagesToReplayNow = messagesToReplay.items(messagesToRead,
```

Comment:
Instead of having the `messagesToReplay ` to be sorted, wouldn't be easier to just make the `messagesToReplayNow` to be a `SortedSet` ? 

In this case, we don't care to have `messagesToReplay` to be *always* sorted, just when there is a redelivery is when we care.

Reply:
>   wouldn't be easier to just make the messagesToReplayNow to be a SortedSet ?

No because `messagesToReplayNow` is a very small sub-set (100 msgs) of `messagesToReplay` (> 1M msgs) and the main issue is  `messagesToReplay` can have more than million random messages and we want to read messages from the same ledger. making ONLY `messagesToReplayNow` sorted set will not help because there is a possibility that 100 read messages from `messagesToReplay` might be random from different ledgers and it will still perform random read across multiple ledgers.

Reply:
Oh that's true.